plutus-core/plutus-core/src/PlutusCore/Core/Type.hs

```diff
@@ -91,6 +92,10 @@ data Term tyname name uni fun ann
     deriving stock (Show, Functor, Generic)
     deriving anyclass (NFData)
 
+-- See Note [ExMemoryUsage instances for non-constants].
+instance ExMemoryUsage (Term tyname name uni fun ann) where
+    memoryUsage = error "Internal error: 'memoryUsage' for Core 'Term' is not supposed to be forced"
```

Comment:
The evaluator doesn't try to get the `ExMemory` of a `Term` as it operates on `CekValue`s, for which we have a proper instance.

> Can we not ensure this isn't needed at compile-time?

No, because it is needed at compile time. The instance is not ignored, it gets picked up and used, it's the thunk that `memoryUsage` creates gets thrown away, because we don't have a builtin whose costing function would actually use the `ExMemory` of an argument whose type is a type variable. We could have such a builtin, it was one of the options for the built-in `show` for example.

I guess my docs are confusing? Any suggestions?

Reply:
Why can't we just... not call it if we don't need it?

Reply:
@michaelpj I don't understand your question. Why can't we express in the types that all polymorphic builtins are sufficiently lazy thus forbidding the possibility of adding builtins that are not lazy, which we may need (it was one of the options for the `show` builtin)? Like how would you even do the first part?

Plus, `TypeScheme` works over a `term`, meaning getting the type of a builtin requires providing some `term` type suitable for evaluation even if you don't want to evaluate anything. I had some progress towards fixing that, but you were explicit about builtins not being a priority, so I closed the [PR](https://github.com/input-output-hk/plutus/pull/4533) and stopped worrying about polishing builtins.

Reply:
I'm operating too far from the details here so sorry if these are silly questions. I indeed would like it if we could somehow make it explicit that this doesn't get called. But it sounds like we can't do that without doing something fancy with the types about laziness, so I guess we have to live with it.

Reply:
> I'm operating too far from the details here so sorry if these are silly questions. I indeed would like it if we could somehow make it explicit that this doesn't get called. But it sounds like we can't do that without doing something fancy with the types about laziness, so I guess we have to live with it.

It's not a big deal if it does get called in the end. It's an instance for `Term`, it's not on a critical path anywhere. We have `error` there just to signal and test that the logic of the builtins is as expected. On the critical path we have a sensible instance for `CekValue` and that one never throws.

Reply:
Right, so the thing that I worry about is being _sure_ that it's not on the critical path. I believe you, it just makes me nervous in case we screw up later...

Reply:
> Right, so the thing that I worry about is being sure that it's not on the critical path. I believe you, it just makes me nervous in case we screw up later...

That has to be a major screwup for us to depend on the `ExMemoryUsage` instance of _typed_ terms on the evaluation path. There's no `ExMemoryUsage` instance for untyped terms:

```
>>> memoryUsage (undefined :: Term () Maybe () ())
<interactive>:36:2: error:
    • No instance for (ExMemoryUsage (Term () Maybe () ()))
        arising from a use of ‘memoryUsage’
```

And thanks to Ziyang we have tests checking that no builtin can throw, which does include costing (because everything on the evaluation path is so strict that there's no way we could not include it).

So much needs to go horribly wrong for us to somehow end up seeing that error triggered on `master` or a release branch.

On the other hand unwillingly defining a weird builtin and then supporting it forever is a very real possibility, so I opted for catching that rather than being worried about things causing synergistically weird behavior (particularly since those things are most likely bugs themselves).

Reply:
> That has to be a major screwup for us to depend on the ExMemoryUsage instance of typed terms on the evaluation path. 

There we go, that's what I was missing lol :facepalm: I thought this was untyped terms.